File: app/src/main/java/ru/p3tr0vich/fuel/ConnectivityHelper.java
package ru.p3tr0vich.fuel;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

public class ConnectivityHelper {

    @Retention(RetentionPolicy.SOURCE)
    @IntDef({DISCONNECTED, CONNECTED, CONNECTED_WIFI, CONNECTED_ROAMING})
    public @interface ConnectedState {
    }

    public static final int DISCONNECTED = 0;
    public static final int CONNECTED = 1;
    public static final int CONNECTED_WIFI = 2;
    public static final int CONNECTED_ROAMING = 3;

    @ConnectedState
    public static int getConnectedState(@NonNull Context context) {
        ConnectivityManager connectivityManager =
                SystemServicesHelper.getConnectivityManager(context);

        if (connectivityManager != null) {
            NetworkInfo activeNetworkInfo = connectivityManager.getActiveNetworkInfo();

            if (activeNetworkInfo != null && activeNetworkInfo.isConnected()) {
                if (activeNetworkInfo.isRoaming())
                    return CONNECTED_ROAMING;

                if (activeNetworkInfo.getType() == ConnectivityManager.TYPE_WIFI)
                    return CONNECTED_WIFI;

                return CONNECTED;
            }
        }

        return DISCONNECTED;
    }
}

File: app/src/main/kotlin/helpers/ConnectivityHelper.kt
package ru.p3tr0vich.fuel.helpers

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.support.annotation.IntDef

object ConnectivityHelper {

    const val DISCONNECTED = 0
    const val CONNECTED = 1
    const val CONNECTED_WIFI = 2
    const val CONNECTED_ROAMING = 3

    @Retention(AnnotationRetention.SOURCE)
    @IntDef(DISCONNECTED, CONNECTED, CONNECTED_WIFI, CONNECTED_ROAMING)
    annotation class ConnectedState

    @JvmStatic
    @ConnectedState
    fun getConnectedState(context: Context): Int {
        val connectivityManager = SystemServicesHelper.getConnectivityManager(context)
                ?: return DISCONNECTED

        val activeNetworkInfo = connectivityManager.activeNetworkInfo ?: return DISCONNECTED

        if (!activeNetworkInfo.isConnected) {
            return DISCONNECTED
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val activeNetwork = connectivityManager.activeNetwork ?: return DISCONNECTED

            val networkCapabilities = connectivityManager.getNetworkCapabilities(activeNetwork)
                    ?: return DISCONNECTED

            if (!networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_ROAMING)) {
                return CONNECTED_ROAMING
            }

            if (networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) {
                return CONNECTED_WIFI
            }

            if (networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_RESTRICTED) &&
                    networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)) {
                return CONNECTED
            }

            return DISCONNECTED
        } else {
            @Suppress("DEPRECATION")
            if (activeNetworkInfo.isRoaming) {
                return CONNECTED_ROAMING
            }

            @Suppress("DEPRECATION")
            if (activeNetworkInfo.type == ConnectivityManager.TYPE_WIFI) {
                return CONNECTED_WIFI
            }

            return CONNECTED
        }
    }
}